rfc/openfed__prerequisite.md

```diff
@@ -0,0 +1,408 @@
+  sku: ID!
+  upc: Int!
+  reviews: [Review!]!
+}
+```
+
+3. Valid: Key field name (`sku`) directly corresponds to a top-level Input field (`ReviewInput.sku`). 
+Key field name `upc` does not directly correspond to a top-level Input field but defines valid mapping 
+(`ReviewInput.productUpc`).
+```graphql
+type Mutation {
+  addReviewToProduct(
+    input: ReviewInput!
+    @openfed__prerequisite(resolveEntity: { typeName: "Product", fields: "sku upc:productUpc" })
+  ): Review
+}
+
+input ReviewInput {
+  content: String!
+  sku: ID!
+  productUpc: Int!
+}
+
+type Review {
+  content: String!
+}
+
+type Product @key(fields: "sku upc") {
+  sku: ID!
+  upc: Int!
+  reviews: [Review!]!
+}
+```
+
+4. Valid: Key field name `upc` directly corresponds to a top-level Input field (`ReviewInput.upc`). 
+Key field name `usku` does not directly correspond to a top-level Input field but defines valid mapping 
+(`ReviewInput.nestedInput.sku`).
+```graphql
+type Mutation {
+  addReviewToProduct(
+    input: ReviewInput!
+    @openfed__prerequisite(resolveEntity: { typeName: "Product", fields: "sku:nestedInput.sku upc" })
+  ): Review
+}
+
+input ReviewInput {
+  content: String!
+  upc: Int!
+  nestedInput: NestedInput!
+}
+
+input NestedInput {
+  sku: ID!
+}
+
+type Review {
+  content: String!
+}
+
+type Product @key(fields: "sku upc") {
+  sku: ID!
+  upc: Int!
+  reviews: [Review!]!
+}
+```
+
+5. Invalid (composition error): Key field name `sku` defines invalid mapping (`ReviewInput.nestedInput.sku` does not 
+exist).
+```graphql
+type Mutation {
+  addReviewToProduct(
+    input: ReviewInput!
+    @openfed__prerequisite(resolveEntity: { typeName: "Product", fields: "sku:nestedInput.sku upc" })
+  ): Review
+}
+
+input ReviewInput {
+  content: String!
+  upc: Int!
+  nestedInput: NestedInput!
+}
+
+input NestedInput {
+  productSku: ID!
+}
+
+type Review {
+  content: String!
+}
+
+type Product @key(fields: "sku upc") {
+  sku: ID!
+  upc: Int!
+  reviews: [Review!]!
+}
+```
+
+6. Invalid (composition error): All key field names define valid mapping but `NestedInput.productSku` of type `[ID]`
+is incompatible with type `ID!`.
+```graphql
+type Mutation {
+  addReviewToProduct(
+    input: ReviewInput!
+    @openfed__prerequisite(resolveEntity: { typeName: "Product", fields: "sku:nestedInput.productSku upc:productUpc" })
+  ): Review
+}
+
+input ReviewInput {
+  content: String!
+  productUpc: Int!
+  nestedInput: NestedInput!
+}
+
+input NestedInput {
+  productSku: Int!
+}
+
+type Review {
+  content: String!
+}
+
+type Product @key(fields: "sku upc") {
+  sku: ID!
+  upc: Int!
+  reviews: [Review!]!
+}
+```
+
+### Entity Representation Call returns data
+If the entity representation call is successful, data will be returned.
+This means that entity reference is valid and the operation cannot proceed.
+
+### Entity Representation Call returns null
+If the entity representation call is unsuccessful, null will be returned.
```

Comment:
Null will be returned by what / to what? In general, I think a better strategy is to pass the error to the subgraph and allow it to read the error and determine how to proceed. Some errors on some fields will be fail open, and others will be fail closed. IMO we should allow the "client" to decide how to handle failures.

Reply:
What if it matches, but some of the fields return errors?

Reply:
I'm not sure what you mean. This is an entity representation call like any other. We just provide indirect access to it through the same subgraph.

Reply:
I think there is some confusion here. This directive validates an entity. It is not perform across-subgraph calls. You send the primary keys and check whether it matches an entity.

The functionality that I think you're describing is something else. This RFC is currently on the sidebench while thinking about what and how we'd like to address injection/inter-subgraph resolutions.

Reply:
Got it. If you do intend to work on this more it would be helpful to add a couple examples showing the actual requests the Router would make to each subgraph including the request body to better understand the call patterns and intent.